### PR TITLE
Improve caching

### PR DESCRIPTION
**Pull request by ps-jay**
Implemented proposed design (from https://trello.com/c/GRfqlKmp):
- If code is not explicitly tagged, then it is built, tested, and then the image is removed
- If code is tagged with a semantic version (vX.Y.Z), then it will:
  - Check if version has been built before:
    - If yes: abort
    - If no: build, test and keep the resulting image
- If code is tagged with anything other than a semantic version, then it will:
  - Check if a build had the same tag previously:
    - If yes: remove the existing image
    - If no: proceed
  - Then proceeds to build, test and keep the resulting image

https://trello.com/c/GRfqlKmp

### CODE REVIEW

File: dockci/models/build.py [hunk 1]
```diff
@@ -352,11 +364,9 @@ def _run_tag_version(self, workdir):
             data_file_path = os.path.join(*stage.data_file_path())
             with open(data_file_path, 'r') as handle:
                 line = handle.readline().strip()
-                # TODO be more generic! GOSH
-                if re.match(r'^v\d+\.\d+\.\d+$', line):
+                if not line == '':
```
Comment (RickyCook) [inline dockci/models/build.py]: `if line != '':`

Comment (RickyCook) [inline dockci/models/build.py]: actually you should be able to just do `if not line:`

File: dockci/models/build.py [hunk 2]
```diff
@@ -392,6 +402,36 @@ def on_done(line):
...
+            existing_image = None
+            for image in self.docker_client.images(
+                name=self.job_slug,
+            ):
+                if tag in image['RepoTags']:
+                    existing_image = image
+                    break
+
+            if existing_image is not None:
+
```
Comment (RickyCook) [inline dockci/models/build.py]: superflous blank line

File: dockci/models/build.py (hunk 1 above)
Comment (ps-jay) [inline dockci/models/build.py]: Done : )

File: dockci/models/build.py [hunk 3]
```diff
@@ -186,6 +187,16 @@ def build_output_details(self):
             if os.path.isfile(path)
         }
 
+    @classmethod
+    def _is_semantic(cls, tag):
```
Comment (RickyCook) [inline dockci/models/build.py]: IMO put this in `docker.utils` as just a fuction

**Pull request merged**